File: lclang/src/main/java/lclang/statements/expressions/AssignExpression.java
package lclang.statements.expressions;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Link;
import lclang.exceptions.LCLangRuntimeException;

public class AssignExpression extends Expression {
    public final Expression left;
    public final Expression right;

    public AssignExpression(Expression left, Expression right) {
        super(0);
        this.left = left;
        this.right = right;
    }

    @Override
    public Link visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        Link value = right.visit(prevCaller, visitor);
        left.visit(prevCaller, visitor).set.invoke(prevCaller, value.get.invoke(prevCaller));
        return value;
    }
}

File: lclang/src/main/kotlin/lclang/statements/AssignExpression.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.Value

class AssignExpression(val left: Expression,
                       val right: Expression,
                       line: Int,
                       column: Int
): Expression(line, column) {

    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        val caller = getCaller(prevCaller)
        val settableValue = left.visit(caller, visitor)
        val value = right.visit(caller, visitor)

        settableValue.set(caller, value)
        return value
    }
}